grub.cfg: switch to new style kernel_status failover handling

anonymouse64 opened the pull request:
This implements the new UC20 spec for loading kernels and failover handling.

Note this is *blocked* until snapd sets the appropriate kernel_status and extracts the kernel assets to ubuntu-boot partition.

The one open question I had about this is what to do if `kernel_status` is not `"try"`, `"trying"`, or `""`. I implemented it such that we default to `""` if `kernel_status` is anything else.

I also renamed grub.cfg-normal to grub.cfg-boot because this grub lives on the ubuntu-boot partition and it's easier to understand which grub corresponds to which partition with this naming scheme I think.

How to test right now:

1. Build an uc20 image with ubuntu-image and a local build of this gadget snap
1. Mount the raw image built by ubuntu-image and add a new ubuntu-boot partition to it, make that a ext4 filesystem
1. Mount the ubuntu-boot partition locally and copy the `grub assets`, `grubx64.efi`, `bootx64.efi` and grub.cfg into `$ubuntu-boot/EFI/boot/grubx64.efi`, `$ubuntu-boot/EFI/boot/bootx64.efi` and `$ubuntu-boot/EFI/ubuntu/grub.cfg` respectively.
1. Generate a grubenv without any variables defined at `$ubuntu-boot/EFI/ubuntu/grubenv` to modify later at runtime
1. Copy a working kernel asset into the ubuntu-boot partition at `$ubuntu-boot/kernel.efi` from the kernel snap
1. Put a "broken" kernel in the ubuntu-boot partition at `$ubuntu-boot/try-kernel.efi` (I just put in a file with "broken" in it there)
1. Unmount the partition and boot the image with qemu (or similar):

```
$ kvm -m 2048 -netdev user,id=mynet0,hostfwd=tcp::8022-:22,hostfwd=tcp::8090-:80 -device virtio-net-pci,netdev=mynet0 -bios /usr/share/OVMF/OVMF_CODE.ms.fd -drive if=virtio,file=uc20-try-kernel.img,format=raw -serial mon:stdio -nographic
```

8. In the first ubuntu-seed grub, choose "Continue into run mode" option
8. Let ubuntu-boot grub boot the default kernel for long enough to see that it is generally working
8. Reboot, and choose "Continue into run mode" option again
8. Enter the grub command line and run the following to confirm that kernel_status is currently unset and then modify it to `"try"` and reboot:

```
grub> load_env --file /EFI/ubuntu/grubenv kernel_status
grub> echo $kernel_status

grub> set kernel_status=try
grub> save_env kernel_status
```
12. Choose "Continue into run mode" option in the ubuntu-seed grub
12. Observe that ubuntu-boot grub fails to boot and reboot the machine:

```
alloc magic is broken at 0x5c807840: 0
Aborted. Press any key to exit.
```
14. Choose "Continue into run mode" option in the ubuntu-seed grub
14. Enter the grub command line and echo `kernel_status` to see it has been reset to `""`:

```
grub> load_env --file /EFI/ubuntu/grubenv kernel_status
grub> echo $kernel_status

grub>
```

anonymouse64 committed:
grub.cfg-normal: rename to grub.cfg-boot

Also update this in the gadget.yaml

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

anonymouse64 committed:
grub.cfg-boot: refactor to use kernel_status and assume kernel/try-kernel

We load bootenv from /EFU/ubuntu/grubenv for this grub on ubuntu-boot partition.
Currently, only kernel_status is stored there, so whitelist just that variable,
and if that variable is anything other than "try", "trying" or "", reset to ""
to proceed with normal boot.

Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

cmatsuoka commented on grub.cfg-boot:
```diff
@@ -0,0 +1,38 @@
+    # we clear the mode and boot normally
+    set kernel_status=""
+    save_env kernel_status
+elif [ x"$kernel_status" != x"" ]; then
```
Is the x technique needed in grub scripts? Or maybe we could simply test for non-emptiness using `-n`?

anonymouse64 replied:
Yes, I wasn't clear on that, I think we might be able to just use `-n`, let me check

xnox replied:
yes use `-n`

xnox commented on grub.cfg-boot:
```diff
@@ -0,0 +1,38 @@
+load_env --file /EFI/ubuntu/grubenv kernel_status
+
+set cmdline="console=ttyS0 console=tty1 panic=-1 systemd.gpt_auto=0 snapd_recovery_mode=run"
+
```
```suggestion
set kernel=kernel.efi
```

xnox commented on grub.cfg-boot:
```diff
@@ -0,0 +1,38 @@
+    # try to boot with try-kernel.img
+    menuentry "Run Ubuntu Core 20" {
+        search --label ubuntu-boot --set=ubuntu_boot
+        chainloader ($ubuntu_boot)/try-kernel.efi $cmdline
```
```suggestion
set kernel=try-kernel.efi
```

xnox commented on grub.cfg-boot:
```diff
@@ -0,0 +1,38 @@
+fi
+
+# boot with the normal kernel.img and initrd.img
+if [ x"$kernel_status" = x"" ]; then
```
```suggestion
```

xnox commented on grub.cfg-boot:
```diff
@@ -0,0 +1,38 @@
+if [ x"$kernel_status" = x"" ]; then
+    menuentry "Run Ubuntu Core 20" {
+        search --label ubuntu-boot --set=ubuntu_boot
+        chainloader ($ubuntu_boot)/kernel.efi $cmdline
```
```suggestion
        chainloader ($ubuntu_boot)/$kernel $cmdline
```

xnox reviewed (commented):
Given that the menuentry should always be disabled, simply toggle the kernel name one is booting, and keep just the single menuentry at the end with dynamic kernel name (either stock, or try one).

I like the grub-boot name. However, I thought we must name the grub.cfg something particular in the gadget. I wish we'd stop shipping 4 grub configs, and limit ourselves to just two with like unique names.

anonymouse64 committed:
grub.cfg-boot: simplify grub menus

Use single menuentry at the end, and store the kernel.efi filename in a
variable for simplicity.

Also use -n instead of the x trick.

Signed-off-by: Ian Johnson <person.uwsome@gmail.com>

anonymouse64 commented:
@xnox done, applied your suggested changes and rebased on 20.

anonymouse64 commented:
Snapd will be ready for this change with https://github.com/snapcore/snapd/pull/7947, so this PR and that one should happen at the same time I think

xnox commented:
@anonymouse64 with a small tweak we can de-couple the two.

One can test for presence of the unpacked kernel, if there isn't one, fallback to the "just boot anything one can find on ubuntu-data" which is what we currently have.
